### PR TITLE
[POC] Cythonizing parts of Dask

### PR DESCRIPTION
**Pull request by shwina**
Experimenting with Cythonizing parts of Dask (see https://github.com/dask/distributed/issues/854 and https://github.com/dask/distributed/issues/854#issuecomment-623734144)

Note: I only opened this PR as a means for facilitating discussion on the above issue - I do not expect for this PR to be considered for merging. 

This PR includes commits from https://github.com/dask/dask/pull/6137

### CODE REVIEW

**Comment by mrocklin** [comment]
This is neat to see.  Thanks @shwina .  

At first glance it looks like the pyx files are the same as the .py files.  Is that so?

If so, I wonder if we could keep them as .py files so that we don't get in the way of scaring off non-Cython developers.  If we're only using Cython lightly, to just compile python files, then hopefully it only ever shows up when we go to build packages, and not during normal development.

I'm also curious what speedups might look like if we do start using Cython more heavily.  There are a variety of tricks one can use which have different performance/readability tradeoffs.

cc @eriknw

**Comment by shwina** [comment]
@mrocklin -- yes, this is basically straight Cythonization of a very small part of Dask without any other changes. If we can add typing information to some of these functions for example, that would help much more.

**Comment by mrocklin** [comment]
> @mrocklin -- yes, this is basically straight Cythonization of a very small part of Dask without any other changes.

For as long as we stay in this mode.  Can we keep the files as `.py` files rather than `.pyx`?  I suspect that that will reduce disruption to the current set of contributors, most of whom are not familiar with Cython.

> If we can add typing information to some of these functions for example, that would help much more.

I would be curious on how much we can improve things here.  My guess is only 1.5-2x, but I would be very glad to learn that we could go further.  I would love to see experimentation here, perhaps starting with some isolated part of the codebase.

Another alternative to Cython is mypyc.  I don't have much familiarity with it, but it might give us the same performance benefits of Cython with the readability and friendliness of pure Python.

**Comment by shwina** [comment]
> For as long as we stay in this mode. Can we keep the files as .py files rather than .pyx? I suspect that that will reduce disruption to the current set of contributors, most of whom are not familiar with Cython.

Sorry -- I don't understand. Are you recommending that we keep the `.py` extension and Python syntax, but compile with Cython? That will preclude adding any possible Cython-specific optimizations going forward. Maybe I am misunderstanding.

Note that I'm not suggesting that this PR ever be considered seriously for review/merge -- it's just a hacky mess to show that Cythonization has some benefits :)

**Comment by jakirkham** [comment]
Yeah both Cython and MyPy have ways to compile based on type annotations. My guess is the benefits would be similar (at the end of the day they are both just using specific Python C APIs as opposed to generic ones). So for simplicity would suggest sticking with Cython since that's what people are familiar with here.

**Comment by shwina** [comment]
> My guess is only 1.5-2x

I'd agree.

That being said, if we replaced the use of e.g., list of strings with `std::vector` or dict with `std::map` within some of these functions, we can expect much more than that.

> Another alternative to Cython is mypyc. I don't have much familiarity with it, but it might give us the same performance benefits of Cython with the readability and friendliness of pure Python.

I haven't used `mypyc` either - it looks really interesting!

**Comment by mrocklin** [comment]
> Note that I'm not suggesting that this PR ever be considered seriously for review/merge -- it's just a hacky mess to show that Cythonization has some benefits :)

Hooray for hacky messes, especially when submitted early :)

> Sorry -- I don't understand. Are you recommending that we keep the .py extension and Python syntax, but compile with Cython? That will preclude adding any possible Cython-specific optimizations going forward. Maybe I am misunderstanding.

Yeah sorry, my original question was intended to be "did you actually use any Cython here, or did you just call the cython compiler on the python code?"  (it's a big diff and so hard to tell when doing a rapid review, my apologies).  It sounds like the answer is "Yes, we used a bit of Cython here, so the `.pyx` extension is important to signal that" which makes sense to me.  

Most of my questions here are trying to maximize performance improvement while minimizing disruption to devs who don't grok Cython.  There is a balance to be struck here.  I'm just trying to explore this space.

> I haven't used mypyc either - it looks really interesting!

Yeah, `mypyc` is the sort of thing that would solve that balance well if it works well for our needs.

**Comment by mrocklin** [comment]
> That being said, if we replaced the use of e.g., list of strings with std::vector or dict with std::map within some of these functions, we can expect much more than that.

I would be *very* curious to see how much we could gain here.

**Comment by shwina** [comment]
> I would be very curious to see how much we could gain here.


Some speedups are reported in https://github.com/dask/distributed/issues/854#issuecomment-622432993 for instance

**Comment by shwina** [comment]
I understand that introducing Cython to the development workflow of Dask/Distributed may add a non-trivial barrier to contribution, especially if Cython proliferates in the code base. A good balance can be struck by limiting Cython to low-level/high-perf data structures and utilities.

Currently, one thing limiting performance is that Dask/Distributed is built on general Python types such as lists/dicts/sets (which is great -- but slow), where it could instead be built on higher performing, **typed** versions of those. That will of course need some large-ish design changes.

**Comment by jakirkham** [comment]
> Currently, one thing limiting performance is that Dask/Distributed is built on general Python types such as lists/dicts/sets (which is great -- but slow), where it could instead be built on higher performing, **typed** versions of those. That will of course need some large-ish design changes.

Is that true? Naively I would assume that if you wrapped up some C++ collections with a bit of Cython that could be an optional external dependency that Dask/Distributed could try to use (falling back to Python objects otherwise). I could be wrong about all of this though.

**Comment by jakirkham** [comment]
[This SO answer]( https://stackoverflow.com/a/29270281 ) seems relevant.

**Comment by shwina** [comment]
@jakirkham agreed. The design changes I mentioned are more about ensuring that Dask objects are composed of more strongly typed collections. For example list of strings, as opposed to lists of arbitratry Python objects.

**Comment by mrocklin** [comment]
> A good balance can be struck by limiting Cython to low-level/high-perf data structures and utilities.

Agreed

> Currently, one thing limiting performance is that Dask/Distributed is built on general Python types such as lists/dicts/sets (which is great -- but slow), where it could instead be built on higher performing, typed versions of those. That will of course need some large-ish design changes.

Agreed.  It's an open question about how much performance we can get before having to change other parts of Dask, like the graphs produced by the various dask.array and dask.dataframe libraries.  

Work in this area though would be very welcome.  We just need to do it in such a way that we're mindful of the tradeoffs.

**Comment by kkraus14** [comment]
May I suggest that we open an issue based on this PR to start exploring / discussing what redesigning / tightening / constraining the task graph and other performance critical interfaces would look like?

I imagine the folks that started looking at the scheduler in Rust would also have a lot of insights / thoughts here as well.

**Comment by mrocklin** [comment]
That sounds good to me!

**Comment by scoder** [comment]
I would also recommend going with `.py` files, and making any (potential future) low-level (C/C++) data structure usage an optional dependency or optional separate `.pyx` module. Letting an `ImportError` make the choice for you is simple enough, but keeping code duplication as small as possible is crucial for future maintenance, especially when non-trivial algorithmic code is involved. Even if you end up rewriting one or two critical helper functions in fine-tuned Cython specific code, making those an optional import is still way better than duplicating hundreds of lines of code.

So far, all of your `.pyx` files are pure Python that just runs faster when Cython compiles it. Why not leave the code where it was and compile the `.py` files instead?

I did the same with nltk as some point, just [compiled the two modules from it that were our main bottlenecks](https://medium.com/trustyou-engineering/speeding-up-the-semantic-analysis-of-guest-reviews-by-30-with-just-one-command-cf66b555260a), and it gave me an overall 30% processing speedup, essentially for free, without otherwise touching the code in nltk. They also ended up [merging this](https://github.com/nltk/nltk/pull/2289), with lots of additional modules compiled, and the only changes were in build scripts.

Basically, what it gives you is the ability to ship a pure Python wheel and additionally compiled, Python version specific binary wheels, from the same implementation. We've been doing this with Cython itself for a very long time.

**Comment by scoder** [comment]
Also, I recommend starting to use Cython 3.0 for this, since that's where the development has been for the last two years or so.

**Comment by jcrist** [comment]
Closing this as stale, if we decide to cythonize some parts of dask later we can reference this PR then.
